File: diagtraverse.py
class Solution:
    def findDiagonalOrder(self, mat: 'List[List[int]]') -> 'List[int]':
        
        if mat is None or len(mat) ==0: return 
        
        m = len(mat)
        n = len(mat[0])
        r,c = 0,0
         
        res = [0]*m*n 
        idx = 0
        dir1 = 1 #going upward dir -1 in downward dir
        
        while(idx < m*n):
            res[idx] = mat[r][c]
            idx += 1
            if (dir1 == 1):
                if c == n-1:
                    dir1 = -1
                    r +=1
                elif r==0:
                    dir1 = -1
                    c +=1
                else:
                    r -=1
                    c +=1
            #dir1 -1 case        
            else:
                if r == m-1:
                    dir1 = 1
                    c += 1
                elif c == 0:
                    dir1 = 1
                    r += 1
                else:
                    c -= 1
                    r += 1
            
        return res

s1 = Solution()
ar = [[1,2,3],[4,5,6],[7,8,9]]
print(s1.findDiagonalOrder(ar))
        

File: DiagTraverse.java
// Time Complexity : O(m*n)
// Space Complexity : O(1) , we dont count output in space complexity
// Did this code successfully run on Leetcode : yes
// Three line explanation of solution in plain english
// Divide the problem into smaller parts and find the edge cases. Then decide the direction and row and col values

// Your code here along with comments explaining your approach

class Solution {
    public int[] findDiagonalOrder(int[][] mat) {
        int m = mat.length; //row
        int n = mat[0].length; //column
        int[] result = new int[m*n];
        int r = 0 ; int c = 0;
        Boolean dir = true; // UP - True & Down --> False

        for( int i = 0 ; i < m*n ; i++){
            // System.out.println("row: " + r);
            // System.out.println("col: " + c);
            result[i] = mat[r][c];
            //up direction
            if(dir){
                if(c == n-1){ // boundary conditions
                    r++;
                    dir = false;

                }else if(r == 0){ // boundary conditions
                    c++;
                    dir = false;
                }
                else{
                    r--;
                    c++;
                }
            }
            else{ //down
                if(c == 0 && r != m-1){ // boundary conditions
                    r++;
                    dir = true;
                }
                else if(r == m-1){ // boundary conditions
                    c++;
                    dir = true;
                }
                else{
                    r++;
                    c--;
                }
            }
        }

        return result;
    }
}